src/clone.rs

```diff
@@ -69,199 +64,187 @@ pub(crate) async fn create_clone<P: Into<Path>>(
+        }
+        None => {
+            let mut parent_manifest =
+                load_initialized_manifest(parent_manifest_store.clone()).await?;
+            let parent_checkpoint =
+                get_or_create_parent_checkpoint(&mut parent_manifest, parent_checkpoint_id).await?;
+            let parent_manifest_at_checkpoint = parent_manifest_store
+                .read_manifest(parent_checkpoint.manifest_id)
+                .await?;
 
-    let final_checkpoint_id = Uuid::new_v4();
-    let parent_db = ParentDb {
-        path: parent_path.to_string(),
-        checkpoint_id: final_checkpoint_id,
+            StoredManifest::create_uninitialized_clone(
+                clone_manifest_store,
+                &parent_manifest_at_checkpoint,
+                parent_path.clone(),
+                parent_checkpoint.id,
+            )
+            .await?
+        }
     };
 
-    let clone_manifest = create_uninitialized_clone_manifest(
-        Arc::clone(&clone_manifest_store),
-        existing_clone_manifest,
-        parent_db,
-        &parent_checkpoint_manifest,
-    )
-    .await?;
-
-    parent_manifest
-        .write_checkpoint(
-            Some(final_checkpoint_id),
-            &CheckpointOptions {
-                lifetime: None,
-                source: Some(parent_checkpoint.id),
-            },
-        )
-        .await?;
-
-    Ok(clone_manifest)
-}
+    // Ensure all external databases contain the final checkpoint.
+    for external_db in &clone_manifest.manifest().external_dbs {
+        let Some(final_checkpoint_id) = external_db.final_checkpoint_id else {
+            // If the final checkpoint id is not set, we can skip this check
+            continue;
+        };
+        let external_db_manifest_store = if external_db.path == parent_path {
+            parent_manifest_store.clone()
+        } else {
+            Arc::new(ManifestStore::new(
+                &external_db.path.clone().into(),
+                object_store.clone(),
+            ))
+        };
+        let mut external_db_manifest =
+            load_initialized_manifest(external_db_manifest_store).await?;
 
-async fn create_uninitialized_clone_manifest(
-    clone_manifest_store: Arc<ManifestStore>,
-    existing_clone_manifest: Option<StoredManifest>,
-    parent_db: ParentDb,
-    parent_manifest: &Manifest,
-) -> Result<StoredManifest, SlateDBError> {
-    match existing_clone_manifest {
-        Some(mut clone_manifest) => {
-            clone_manifest
-                .rewrite_parent_db(parent_db, parent_manifest)
+        if external_db_manifest
+            .db_state()
+            .find_checkpoint(&final_checkpoint_id)
+            .is_none()
+        {
+            external_db_manifest
+                .write_checkpoint(
+                    Some(final_checkpoint_id),
+                    &CheckpointOptions {
+                        lifetime: None,
+                        source: Some(external_db.source_checkpoint_id),
```

Comment:
I think it's possible for this checkpoint `external_db.source_checkpoint_id` to be expired/deleted at this point (in the case where we fail and then retry this fn later). Can we add a check for that and fail?

Reply:
There are two cases when source checkpoint no longer exists:
- Final Checkpoint already exists - this is a valid state and there is no need to fail
- Neither source nor final checkpoint exists - in this case we already fail with CheckpointNotFound

Here is a test case -> https://github.com/slatedb/slatedb/pull/514

Did you have something else in mind?

Reply:
The case I was thinking of:

1. call create clone with `parent_checkpoint: None`
2. `create_clone_manifest` creates an ephemeral checkpoint C_p in the parent db, and sets C_p as the `source_checkpoint_id` in the `external_dbs` for the parent.
3. FAIL
4. C_p expires, and the GC deletes it (and potentially some SSTs it references)
5. call create clone with `parent_checkpoint: None`
6. `create_clone_manifest` creates the checkpoint from expired C_p <-- This actually can't succeed because the implementation of `write_checkpoint` will look for C_p and fail. I forgot about the check in `write_checkpoint`.

So nvm - this should actually be fine.